File: go/fixture/go-mini/internal/services/validate.go
package services

import (
	"errors"
	"fmt"
	"strings"

	"example.com/go-mini/internal/models"
)

// validRecipient reports whether the alert's recipient fits its channel.
func validRecipient(a models.Alert) bool {
	//nolint:goconst // TODO
	switch a.Channel {
	case "email":
		return strings.Contains(a.Recipient, "@")
	case "slack":
		return strings.HasPrefix(a.Recipient, "#")
	}
	return false
}

func ValidateAlert(a models.Alert) error {
	if a.Channel == "" {
		return errors.New("alert: empty channel")
	}
	if a.Summary == "" {
		return errors.New("alert: empty summary")
	}
	if len(a.Summary) > 512 {
		return fmt.Errorf("alert: summary of %d bytes is too long", len(a.Summary))
	}
	if !validRecipient(a) {
		return fmt.Errorf("alert: bad recipient %q for channel %q", a.Recipient, a.Channel)
	}
	return nil
}

// Deliver validates an alert and sends it.
func Deliver(a models.Alert) error {
	if err := ValidateAlert(a); err != nil {
		return err
	}
	return Send(a)
}

File: py/fixture/py-mini/internal/services/validate.py
"""Alert validation before delivery."""

from internal.models.alert import Alert
from internal.services.notify import send_alert


def _valid_recipient(a: Alert) -> bool:
    """Report whether the alert's recipient fits its channel."""
    match a.channel:
        case "email":
            return "@" in a.recipient  # noqa: PLR2004  # TODO
        case "slack":
            return a.recipient.startswith("#")
    return False


def validate_alert(a: Alert) -> None:  # noqa: D103
    if not a.channel:
        raise ValueError("alert: empty channel")
    if not a.summary:
        raise ValueError("alert: empty summary")
    if len(a.summary) > 512:
        raise ValueError(f"alert: summary of {len(a.summary)} bytes is too long")
    if not _valid_recipient(a):
        raise ValueError(f'alert: bad recipient "{a.recipient}" for channel "{a.channel}"')


def deliver(a: Alert) -> None:
    """Validate an alert and send it."""
    validate_alert(a)
    send_alert(a)
